design/baremetal-operator/bmc-events.md

```diff
@@ -0,0 +1,139 @@
+subscription. While Redfish is the primary target for this design, the
+Ironic API is vendor-neutral and seeks to provide a unified interface
+for configuring events.
+
+## Motivation
+
+Some environments run workloads that need to deal with potential
+faults or environmental changes quicker than they would get an alert
+through other channels. For example, some workloads may have a sidecar
+container that knows how to deal with an alert that a particular network
+interface went down, or that the CPU temperature reached a certain
+threshold.
+
+### Goals
+
+- Provide an API to manage subscriptions to events
+
+### Non-Goals
+
+- [Configurable events and thresholds](#configurable-events-and-thresholds)
+- Any kind of event polling
+- Software for processing the events, i.e. any webhook
+- BMC's beyond Redfish for now
+
+## Proposal
+
+### User Stories
+
+- As a user, I'd like to configure a my BMC to send events to a target URL.
+- As a user, I'd like to filter the types of events the BMC sends to my target URL.
+- As a user, I'd like to provide context to a particular event subscription.
+- As a user, I'd like to provide arbitrary headers.
+
+## Design Details
+
+### Implementation Details
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BMCEventSubscription
+metadata:
+  name: worker-1-events
+spec:
+   hostRef: ostest-worker-1
+   targetURI: https://events.apps.corp.example.com/webhook
+   filters:
+     - StatusChange
+     - ResourceAdded
+     - Alert
+   headers:
+     X-Event-Source: "metal3.io"
+   headerRef: webhookBridgeAuth
+   context: “SomeUserContext”
+status:
+  errorMessage: ""
+  errorCount: 0
+  subscriptionID: aa618a32-9335-42bc-a04b-20ddeed13ade
+```
+
+- A BMCEventSubscription resource represents a subscription to the events generated
+  by a specific BMC.
+- Ironic will manage configuring the subscription, using a new API for managing them.
+- The BMCEventSubscription with maintain a reference to a BareMetalHost.
+- The BMCEventSubscription will maintain a reference to the ironic
+  subscription ID.
+- The BMCEventSubscription will allow injection of headers inline, or
+  using a headerRef to a secret, for example to provide basic auth
+  credentials.
+
+### Open Questions
+
+#### Configurable events and thresholds
+
+Can we support configurable events and thresholds?
+
+Redfish standard itself does not seem to have a way to specify specific
+alerts and thresholds. For example, to receive an alert when the
+temperature exceeds 40C, one would need to configure this manually
+according to the vendor's reccomendations.
+
+Vendors, however, do provide vendor-specific ways to configure these
+thresholds, but it's hard to abstract to a neutral interface. For
+example, here is a [Dell example for temperature](https://www.dell.com/support/manuals/en-jm/idrac9-lifecycle-controller-v4.x-series/idrac9_4.00.00.00_redfishapiguide_pub/temperature?guid=guid-5a798111-407b-485d-b6fb-7d6e367d4ad4&lang=en-us).
```

Comment:
@dtantsur @iurygregory Do you all have any idea if this is possible? I discussed with some folks who are interested in this feature, and they are also interested in configuring thresholds.  My initial research indicates this is highly vendor specific, if it's available at all (Supermicro docs make no reference, Dell has some of it's own API's to do this)

Reply:
Do we need to support configuring the thresholds to make the API useful?

Reply:
The API is useful by itself to configure the target URI for the webhook, but it'd be more useful if we could configure thresholds. Otherwise it has to be done out of band, or users accept whatever the vendors default alerts are.

Reply:
If we're going to configure limits in the same API where we register the web hook, does that imply that we only want to support 1 event type at a time in this API? What happens if 2 hooks are registered for the same event with different limits? Maybe the limits live on the host, instead of the event subscription?

Reply:
@stbenjam  this would be very vendor specific,  also looking the example from Dell this is not under the Events so it will probably be a different API in Ironic to support this.

Reply:
I agree with @dhellmann , the limits would be something for the host (and in the dell example we can see the endpoint has nothing to do with EventSubscription). Also, the subscription API in redfish can receive a list of event types, I don't think it would make sense to limit this on Ironic because users want to configure thresholds.

Reply:
> If we're going to configure limits in the same API where we register the web hook, does that imply that we only want to support 1 event type at a time in this API? What happens if 2 hooks are registered for the same event with different limits? Maybe the limits live on the host, instead of the event subscription?

I think this would be a separate API, for configuring an event type.  Because you would be configuring a TemperatureOver40C alert, and multiple hosts subscribing to it.

Although I think my main point is being missed here, I don't think this is possible and I'm explicitly calling it out of scope for this Metal3 API unless someone thinks it's possible and I'm misreading the Redfish docs.

Reply:
++ for a non-goal for this enhancement

Reply:
Ok I'll leave it as a non-goal, take this out of open questions, but at least talk about it a bit in Alternatives.